modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/set/IgniteMapMinus.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.ignite.internal.processors.query.calcite.trait.TraitUtils;
+import org.apache.ignite.internal.processors.query.calcite.type.IgniteTypeFactory;
+import org.apache.ignite.internal.processors.query.calcite.util.Commons;
+
+/**
+ * Physical node for MAP phase of MINUS (EXCEPT) operator.
+ */
+public class IgniteMapMinus extends IgniteMinusBase {
+    /** */
+    public IgniteMapMinus(
+        RelOptCluster cluster,
+        RelTraitSet traitSet,
+        List<RelNode> inputs,
+        boolean all
+    ) {
+        super(cluster, traitSet, inputs, all);
+    }
+
+    /** */
+    public IgniteMapMinus(RelInput input) {
+        super(input);
+    }
+
+    /** {@inheritDoc} */
+    @Override public IgniteMapMinus copy(RelTraitSet traitSet, List<RelNode> inputs, boolean all) {
+        return new IgniteMapMinus(getCluster(), traitSet, inputs, all);
+    }
+
+    /** {@inheritDoc} */
+    @Override public IgniteRel clone(RelOptCluster cluster, List<IgniteRel> inputs) {
+        return new IgniteMapMinus(cluster, getTraitSet(), Commons.cast(inputs), all);
+    }
+
+    /** {@inheritDoc} */
+    @Override public <T> T accept(IgniteRelVisitor<T> visitor) {
+        return visitor.visit(this);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected RelDataType deriveRowType() {
+        RelDataTypeFactory typeFactory = Commons.typeFactory(getCluster());
+
+        assert typeFactory instanceof IgniteTypeFactory;
+
+        RelDataTypeFactory.Builder builder = new RelDataTypeFactory.Builder(typeFactory);
+
+        builder.add("GROUP_KEY", typeFactory.createJavaType(GroupKey.class));
+        builder.add("COUNTERS", typeFactory.createJavaType(int[].class));
+
+        return builder.build();
+    }
+
+    /** {@inheritDoc} */
+    @Override public List<Pair<RelTraitSet, List<RelTraitSet>>> deriveRewindability(
+        RelTraitSet nodeTraits,
+        List<RelTraitSet> inputTraits
+    ) {
+        boolean rewindable = inputTraits.stream()
+            .map(TraitUtils::rewindability)
+            .allMatch(RewindabilityTrait::rewindable);
+
+        if (rewindable)
+            return ImmutableList.of(Pair.of(nodeTraits.replace(RewindabilityTrait.REWINDABLE), inputTraits));
+
+        return ImmutableList.of(Pair.of(nodeTraits.replace(RewindabilityTrait.ONE_WAY),
+            Commons.transform(inputTraits, t -> t.replace(RewindabilityTrait.ONE_WAY))));
+    }
+
+    /** {@inheritDoc} */
+    @Override public List<Pair<RelTraitSet, List<RelTraitSet>>> deriveDistribution(
+        RelTraitSet nodeTraits,
+        List<RelTraitSet> inputTraits
+    ) {
+        Set<IgniteDistribution> distributions = inputTraits.stream()
+            .map(TraitUtils::distribution)
+            .collect(Collectors.toSet());
+
+        ImmutableList.Builder<Pair<RelTraitSet, List<RelTraitSet>>> b = ImmutableList.builder();
+
+        for (IgniteDistribution distribution : distributions) {
```

Comment:
In this case broadcast input can be planned on the MAP phase to processed together with random input on each node, which produces wrong results. `CalciteQueryProcessorTest#testExceptMerge` reproduces this issue.

Reply:
Yes, you're right. The rule should be a little more complex:
```
        return ImmutableList.of(
            Pair.of(nodeTraits.replace(IgniteDistributions.random()), Commons.transform(inputTraits,
                t -> t.replace(t.getDistribution() == IgniteDistributions.single() ? IgniteDistributions.single() : IgniteDistributions.random())
            ))
        );
```
Perhaps, it worth to add additional planner's test to verifying a plan with bunch of table of all spectrum of distribution (affinity, single, broadcast)

The problem with current approach is that it will create unnecessary nodes. Assume we have N inputs with N different hash distributions. After deriving we will have N different variant for each distribution. This then creates N * (N - 1) converters (if possible) for every variant, whereas we could simple do all the job in single map-reduce phase.

Reply:
Fixed.